File: src/lib/ipc/picus/plan.ts
/**
 * Picus query plans — `EXPLAIN`, and the separate, explicit `EXPLAIN ANALYZE`.
 *
 * Its own module rather than another section of `db.ts` because it is its own
 * thing: everything here is about a statement the user is **not** running, which
 * is the opposite of what every call in `db.ts` does — with the one exception that
 * makes this dangerous and is therefore named in every doc comment below.
 *
 * ## Estimate versus measurement
 *
 * Without `analyze` the server plans the statement and executes nothing: every
 * number is a guess. With `analyze` the server **runs the statement** and reports
 * what actually happened. The two are not display options of one another, and the
 * answer carries {@link QueryPlan.analyzed} so nothing downstream has to remember
 * which it asked for.
 *
 * The backend refuses `analyze` for anything that is not a read, and on a read-only
 * connection refuses it as a write. That refusal is the engine's, not this layer's:
 * a client-side guess about what counts as a write would only ever be a second,
 * weaker opinion.
 */

import { picus } from '../rpc';

/** One step of the plan. Flat, with {@link depth} carrying the tree. */
export interface PlanNode {
  /** Indentation depth; 0 is the root. */
  depth: number;
  /** `Seq Scan`, `Index Scan using orders_pkey`, `Hash Join`, … */
  label: string;
  /** The relation this node reads, when it reads one. */
  relation: string | null;
  /** Estimated total cost, in the engine's own units. */
  cost: number | null;
  /** Estimated rows out — **per loop**, like `actualRows` beside it. */
  rows: number | null;
  /** Rows actually produced. Only present on an analysed plan. */
  actualRows: number | null;
  /** Time actually spent. Only present on an analysed plan. */
  actualMs: number | null;
  /** Filters, sort keys, index conditions, buffer accounting — the server's words. */
  detail: string[];
  /** A remark worth surfacing, in prose. Absent far more often than present. */
  warning: string | null;
}

/** A plan, as text and as structure. */
export interface QueryPlan {
  /** The engine's own output, verbatim — what gets pasted into a ticket. */
  text: string;
  /** The same plan, one entry per node, in execution-tree order. */
  nodes: PlanNode[];
  /**
   * The statement was executed. **`false` means every number here is an estimate**,
   * and the interface must say so — this is the single most important field on the
   * screen it draws.
   */
  analyzed: boolean;
  /** Total estimated cost of the root node. */
  totalCost: number | null;
  /** Wall time of the whole plan, when it was analysed. */
  actualMs: number | null;
  /** How long producing the plan took. */
  elapsedMs: number;
}

/**
 * Ask for a statement's plan.
 *
 * `analyze` **runs the statement**. Never send it because a plan would be more
 * useful with real numbers in it — send it because the user asked for it, on a
 * statement they are willing to have executed. Omitted, it is false.
 */
export function explainQuery(
  connectionId: string,
  sql: string,
  analyze = false,
): Promise<QueryPlan> {
  return picus('picus_explain', { connectionId, sql, analyze });
}

File: crates/products/picus/db-provider/api/src/plan.rs
//! The query plan — what the server says it will do, or what it just did.
//!
//! Carried as a **tree flattened to a depth-tagged list** rather than as nested
//! nodes. Two reasons, and the second is the real one: the interface renders it as
//! an indented list either way, and a flat list serialises across the RPC seam
//! without a recursive type that every consumer then has to walk.
//!
//! ## `ANALYZE` is not a display option
//!
//! `EXPLAIN` describes; `EXPLAIN ANALYZE` **runs the statement**. On a `SELECT`
//! that is merely slow; on a `DELETE` it is the delete. So it is a separate,
//! explicit request ([`PlanRequest::analyze`]), an implementation must refuse it
//! for anything that is not a read, and the answer says which of the two it is
//! ([`QueryPlan::analyzed`]) so the interface never labels an estimate as a
//! measurement.

use serde::{Deserialize, Serialize};

/// What to explain, and how.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanRequest {
    /// Actually run the statement and report real times and row counts.
    ///
    /// Must be refused for anything that is not a read, and on a read-only
    /// connection is the only form that can be refused at all — which is the point:
    /// the refusal belongs in the engine, where it cannot be bypassed.
    pub analyze: bool,
    /// Include buffer accounting where the engine has it.
    pub buffers: bool,
}

/// A plan, as text and as structure.
///
/// Both, deliberately. The structure is what the interface draws; the text is what
/// gets pasted into a ticket, and a tool that can only show you its own rendering
/// of a plan is a tool you cannot ask anyone else about.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct QueryPlan {
    /// The engine's own textual plan, verbatim.
    pub text: String,
    /// The same plan, one entry per node, in execution-tree order.
    pub nodes: Vec<PlanNode>,
    /// The statement was executed. `false` means every number below is an estimate.
    pub analyzed: bool,
    /// Total estimated cost of the root node, in the engine's own units.
    pub total_cost: Option<f64>,
    /// Wall time of the whole plan when it was analysed.
    pub actual_ms: Option<f64>,
    /// How long producing the plan took.
    pub elapsed_ms: u64,
}

/// One step of the plan.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct PlanNode {
    /// Indentation depth; 0 is the root.
    pub depth: u32,
    /// `Seq Scan`, `Index Scan using …`, `Hash Join`, …
    pub label: String,
    /// The relation this node reads, when it reads one.
    pub relation: Option<String>,
    /// Estimated total cost.
    pub cost: Option<f64>,
    /// Estimated rows out.
    pub rows: Option<f64>,
    /// Rows actually produced — only when analysed.
    pub actual_rows: Option<f64>,
    /// Time actually spent — only when analysed.
    pub actual_ms: Option<f64>,
    /// Extra lines the engine attached: filters, sort keys, index conditions.
    pub detail: Vec<String>,
    /// A remark worth surfacing — a sequential scan of a large relation, an
    /// estimate that missed by an order of magnitude. Stated as prose because it
    /// is advice, and advice with no reasoning attached is noise.
    pub warning: Option<String>,
}
